### PR TITLE
ENH: linalg: speedup _sqrtm_triu by moving tight loop to Cython

### PR DESCRIPTION
**Pull request by pv**
#### Reference issue
See gh-12464

#### What does this implement/fix?
Speed up `sqrtm` and `logm` by moving a tight loop in `_sqrtm_triu` to Cython. After this change, most time in `_sqrtm_triu` is spent in D/ZTRSYL (and for the total algorithm, schur and various other parts starts to dominate), so Cython won't help for speeding up the remaining part.

Also add benchmarks.

#### Additional information
Benchmark results:
```
$ python3 runtests.py --bench-compare master linalg_sqrtm linalg_logm
...
       before           after         ratio
     [fb7127c5]       [a34ecce7]
     <master>         <sqrtm-triu-speedup>
-         250±1ms          236±1ms     0.94  linalg_sqrtm.Sqrtm.time_sqrtm('complex128', 256, 32)
-         266±1ms          235±1ms     0.89  linalg_sqrtm.Sqrtm.time_sqrtm('complex128', 256, 64)
-       118±0.5ms        102±0.6ms     0.87  linalg_sqrtm.Sqrtm.time_sqrtm('float64', 256, 32)
-       132±0.1ms          103±1ms     0.78  linalg_sqrtm.Sqrtm.time_sqrtm('float64', 256, 64)
-         663±6ms          484±3ms     0.73  linalg_logm.Logm.time_logm('complex128', 256, 'gen')
-       547±0.6ms        398±0.6ms     0.73  linalg_logm.Logm.time_logm('complex128', 256, 'her')
-         558±2ms          378±2ms     0.68  linalg_logm.Logm.time_logm('complex128', 256, 'pos')
-         359±2ms        237±0.4ms     0.66  linalg_sqrtm.Sqrtm.time_sqrtm('complex128', 256, 256)
-         513±4ms          337±8ms     0.66  linalg_logm.Logm.time_logm('float64', 256, 'gen')
-         415±9ms          265±1ms     0.64  linalg_logm.Logm.time_logm('float64', 256, 'her')
-     10.0±0.05ms      6.38±0.03ms     0.64  linalg_sqrtm.Sqrtm.time_sqrtm('complex128', 64, 32)
-     9.45±0.08ms      5.79±0.08ms     0.61  linalg_sqrtm.Sqrtm.time_sqrtm('float64', 64, 32)
-     13.8±0.08ms      6.31±0.05ms     0.46  linalg_sqrtm.Sqrtm.time_sqrtm('complex128', 64, 64)
-       225±0.5ms        102±0.6ms     0.45  linalg_sqrtm.Sqrtm.time_sqrtm('float64', 256, 256)
-     13.2±0.09ms      5.70±0.08ms     0.43  linalg_sqrtm.Sqrtm.time_sqrtm('float64', 64, 64)
-         275±1ms          102±1ms     0.37  linalg_logm.Logm.time_logm('float64', 256, 'pos')
-      59.0±0.9ms       20.2±0.1ms     0.34  linalg_logm.Logm.time_logm('complex128', 64, 'gen')
-      53.4±0.3ms       16.3±0.4ms     0.31  linalg_logm.Logm.time_logm('float64', 64, 'gen')
-      53.5±0.4ms       16.3±0.1ms     0.30  linalg_logm.Logm.time_logm('complex128', 64, 'pos')
-      52.8±0.2ms       15.3±0.2ms     0.29  linalg_logm.Logm.time_logm('complex128', 64, 'her')
-      46.9±0.4ms      9.29±0.09ms     0.20  linalg_logm.Logm.time_logm('float64', 64, 'her')
-      53.1±0.7ms      9.91±0.03ms     0.19  linalg_logm.Logm.time_logm('float64', 64, 'pos')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

### CODE REVIEW

File: scipy/linalg/_matfuncs_sqrtm_triu.pyx [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# cython: boundscheck=False, wraparound=False, cdivision=True
+import numpy as np
+from ._matfuncs_sqrtm import SqrtmError
+
+from numpy cimport complex128_t, float64_t, intp_t
+
+
+cdef fused floating:
```
Comment (ilayn) [inline scipy/linalg/_matfuncs_sqrtm_triu.pyx:8]: I haven't checked but are we expecting any esoteric numpy types e.g. int16 etc. here?

Comment (pv) [reply scipy/linalg/_matfuncs_sqrtm_triu.pyx:8]: The code in `_sqrtm_triu` casts the input to either of these two

**Review by ilayn — APPROVED** [review]

**Comment by ilayn** [comment]
Looks good to me with a nice speedup. +1 for merge.

**Comment by ev-br** [comment]
LGTM too. Merged, thanks Pauli, Ilhan

**Pull request merged**